pkg/operator/noobaa/operator.go

```diff
@@ -0,0 +1,294 @@
+func (o *Operator) GetControllerOwner(obj interface{}) (*meta_v1.OwnerReference, meta_v1.Object) {
+	metaObject, err := meta.Accessor(obj)
+	if err != nil {
+		deleted, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			logger.Errorf("Error decoding object, invalid type %+v", reflect.TypeOf(obj))
+			return nil, nil
+		}
+		metaObject, err := meta.Accessor(deleted.Obj)
+		if err != nil {
+			logger.Errorf("Error decoding deleted object, invalid type")
+			return nil, nil
+		}
+		logger.Infof("Recovered deleted object: %s", metaObject.GetSelfLink())
+	}
+	owner := meta_v1.GetControllerOf(metaObject)
+	return owner, metaObject
+}
+
+// NewObjectEventHandler returns a watch handler based on a simple handler func
+// with general handling of updates to same resource version.
+func (o *Operator) NewObjectEventHandler(objectHandler func(obj interface{})) cache.ResourceEventHandlerFuncs {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    objectHandler,
+		DeleteFunc: objectHandler,
+		UpdateFunc: func(old, obj interface{}) {
+			newMeta, err := meta.Accessor(obj)
+			if err != nil {
+				logger.Tracef("Operator: Ignore non meta object event %+v", reflect.TypeOf(obj))
+				return
+			}
+			oldMeta, err := meta.Accessor(old)
+			if err != nil {
+				logger.Tracef("Operator: Ignore non meta old object event %+v", reflect.TypeOf(old))
+				return
+			}
+			if newMeta.GetResourceVersion() == oldMeta.GetResourceVersion() {
+				logger.Tracef("Operator: Ignore cached ResourceVersion=%s for %s",
+					newMeta.GetResourceVersion(), newMeta.GetSelfLink())
+				return
+			}
+			objectHandler(obj)
+		},
+	}
+}
+
+// MustCreateCRDFromFile reads a yaml/json file with CRD spec and creates it if it doesn't exist.
+// It will retry while respecting the operator's stop channel close events.
+// Stopping here is probably less relevant for a running operator, mostly for testing.
+func (o *Operator) MustCreateCRDFromFile(objType runtime.Object, crdFileName string) {
+	crd := &ext_v1.CustomResourceDefinition{}
+	crdFilePath := crdFileName
+	if o.DevMode {
+		crdFilePath = "cluster/examples/kubernetes/noobaa/" + crdFileName
+	}
+	err := ReadObjectFromFile(crdFilePath, crd)
+	rook.TerminateOnError(err, "Operator: Failed loading resource from file "+crdFilePath)
+
+	err = wait.PollImmediateUntil(time.Second, func() (bool, error) { return o.CreateCRD(crd) }, o.StopChan)
+	rook.TerminateOnError(err, fmt.Sprintf("Operator: %s CRD could not be installed", crd.Kind))
+
+}
+
+// CreateCRD uses APIExtClient to create the CRD if it doesn't exist.
+// Returns true if created/exists, false to retry, error on unrecoverable error.
+func (o *Operator) CreateCRD(crd *ext_v1.CustomResourceDefinition) (bool, error) {
+	crdClient := o.APIExtClient.ApiextensionsV1beta1().CustomResourceDefinitions()
+	crdReply, err := crdClient.Get(crd.GetName(), meta_v1.GetOptions{})
+	if errors.IsNotFound(err) {
+		_, err = crdClient.Create(crd)
```

Comment:
@travisn I saw you wrote that rook operators don't have permission to create CRDs.
So far I was running the operator with the permissions, but I can make it optional so it will try to create once or so.
What do you think?

Reply:
Long ago the rook-ceph operator created all the resources automatically it needed... the CRDs, service accounts, etc. This required the operator to have full privileges for the lifetime of the operator, which is indefinitely. After a security review, we moved to minimal privileges where we define only the privileges absolutely necessary with the RBAC. This ensures that the long-running operator doesn't have admin privs. The admin can use their higher privs once to create the CRDs, service accounts, etc, rather than requiring the operator to have them. It isn't as clean of a UX since all the yaml must be creaed separately by the admin, but it's a much better security approach. See [common.yaml](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/common.yaml) for rook-ceph's RBAC, though it will probably be much more complex than what Noobaa needs.

Reply:
OK thanks, will follow the same.